url8012c8d031_dmbb966_BBMS_tgz-pJ8-testbed_JNEATtestJ8/src/hex/Hex.java

```diff
@@ -0,0 +1,552 @@
+			GlobalFuncs.scenMap.vaporSourceList.remove(this);
+			GlobalFuncs.ticksStable = 0;			// Removing a source shouldn't lead to a huge DV, but just in case.
+		}
+		
+		vaporType = VaporEnum.NONE;
+	}
+		
+	/**
+	 * Goes through the calculations that CalcVapor does, but stores them in temporary variables and returns it
+	 * @return
+	 */
+	public int ReturnVaporCalc() {
+		int vOut = 0;
+				
+		for (int direction = 0; direction < 6; direction++) {
+			HexOff target = new HexOff(x, y).findNeighbor(direction);
+			Hex tgtHex = GlobalFuncs.scenMap.getHex(target.x, target.y);			
+			int vaporXferAmount = CalcVaporOneHex(this, tgtHex);			
+			
+			// We don't need to worry about restricting updates to positive responses since we are only looking
+			// at a single hex in isolation (whereas when doing a map update, the surrounding hexes will
+			// add equal and opposite flows.  You know - skew symmetry, or something like that.
+			vOut += vaporXferAmount;		 							
+		}		
+		
+		return Math.abs(vOut);
+	}
+	
+	/**
+	 * Calculates the vapor transfer out of the origin hex to the target
+	 * @param origin
+	 * @param target
+	 * @return
+	 */
+	public int CalcVaporOneHex(Hex origin, Hex tgtHex) {
+		int vOut = 0;
+		
+		if (tgtHex.tEnum != TerrainEnum.INVALID) {
+			int moveCost = Math.max(this.tType.getMoveCost(MoveClass.TRACK), 
+					tgtHex.tType.getMoveCost(MoveClass.TRACK));
+			double vaporXferRate = (0.16 * GlobalFuncs.flowRate) / moveCost;
+			int dVapor = origin.vapor - tgtHex.vapor;				
+			
+			vOut = (int)(dVapor * vaporXferRate); 
+		}
+		return vOut;
+	}
+	
+	/**
+	 * DEFAULTS to MoveClass.TRACK
+	 */
+	public void CalcVapor() {
+		for (int direction = 0; direction < 6; direction++) {
+			HexOff target = new HexOff(x, y).findNeighbor(direction);
+			Hex tgtHex = GlobalFuncs.scenMap.getHex(target.x, target.y);
+			
+			int vaporXferAmount = CalcVaporOneHex(this, tgtHex);
+			
+			if (vaporXferAmount > 0) {
+				vaporOut += vaporXferAmount;
+				tgtHex.vaporIn += vaporXferAmount;
+			}
+								
+		}
+	}
+	
+	public double CalcSharedDV() {
+		return (double)deltaVapor / numSpots;
+	}
+	
+	/**
+	 * Predicts the equilibrium vapor level of this hex using a heuristic
+	 * (linear interpolation between the nearest source and sink).
+	 * Must have both a source and a sink on the map; otherwise it does nothing.
+	 */
+	public void PredictVaporLevel() {
+		if (GlobalFuncs.scenMap.vaporSourceList.size() == 0 || GlobalFuncs.scenMap.vaporSinkList.size() == 0) return;
+		if (vaporType != VaporEnum.NONE) return;
+		
+		int sourceDist = 9999;
+		int sinkDist = 9999;
+		HexOff myLoc = new HexOff(this.x, this.y);
+		
+		// Find the nearest source
+		for (int i = 0; i < GlobalFuncs.scenMap.vaporSourceList.size(); i++) {
+			Hex finger = GlobalFuncs.scenMap.vaporSourceList.elementAt(i);
+			HexOff locSrc = new HexOff(finger.x, finger.y);
+			int distance = myLoc.DistFrom(locSrc);
+			
+			if (distance < sourceDist) sourceDist = distance;
+		}
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 281:287, 290:296.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.